dimod/include/dimod/quadratic_model.h

```diff
@@ -0,0 +1,576 @@
+// Copyright 2021 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+namespace dimod {
+
+enum Vartype { BINARY, SPIN, INTEGER };
+
+template <class Bias, class Neighbor>
+class Neighborhood {
+ public:
+    using bias_type = Bias;
+    using neighbor_type = Neighbor;
+    using size_type = std::size_t;
+
+    struct iterator {
+        // this could be random access but let's keep it simple for now
+        using iterator_category = std::forward_iterator_tag;
+        using difference_type = std::ptrdiff_t;
+        using value_type = std::pair<const neighbor_type&, bias_type&>;
+        using pointer = value_type*;
+        using reference = value_type;
+
+        using bias_iterator = typename std::vector<bias_type>::iterator;
+        using neighbor_iterator =
+                typename std::vector<neighbor_type>::const_iterator;
+
+        iterator(neighbor_iterator nit, bias_iterator bit)
+                : neighbor_it_(nit), bias_it_(bit) {}
+
+        iterator& operator++() {
+            bias_it_++;
+            neighbor_it_++;
+            return *this;
+        }
+
+        iterator operator++(int) {
+            iterator tmp = *this;
+            ++(*this);
+            return tmp;
+        }
+
+        friend bool operator==(const iterator& a, const iterator& b) {
+            return a.neighbor_it_ == b.neighbor_it_ && a.bias_it_ == b.bias_it_;
+        }
+
+        friend bool operator!=(const iterator& a, const iterator& b) {
+            return a.neighbor_it_ != b.neighbor_it_ && a.bias_it_ != b.bias_it_;
+        }
+
+        value_type operator*() { return value_type{*neighbor_it_, *bias_it_}; }
+
+     private:
+        bias_iterator bias_it_;
```

Comment:
In c++ we follow [google's style guide](https://google.github.io/styleguide/cppguide.html#Variable_Names). In python you're right, [the convention is to use a leading `_`](https://www.python.org/dev/peps/pep-0008/#id36). And cython who knows :laughing:

Reply:
Once an arguments name was say name, and the member was name_,  i used name instead of name_, should not have happened if i had used _name instead of name_, this is an isolated rare case though :)